readme.txt
Decker (Java port)
version : 77%



NOT PLAYABLE YET



This is the unfinished Java port of the game Decker.
The road map lists the unfinished parts.
Requires Sun's JRE 5 to run. Get it from
http://java.com


See the decker project page for more information about the game :
http://sourceforge.net/projects/decker/


The port is not 1:1.
The main differences between Decker 1.9 and Decker 1.9.1 (Java port) will be :
- the game now has a scripting language for missions
- most message boxes have been turned into tooltips
- it is possible to have several instances of the same program running
  at the same time during a run
- support of custom decker sprites
- localization (english and german will be the first localizations)
- 100% Pure Java, so the program should run on almost every PC
- it is possible to accept several missions at the same time and to drop them again

- the level of the offered missions is based on the levels of the missions
  the player completes or fails. reputation and lifestyle are only used for the
  upper limit of the mission ratings for generic missions. lifestyle still
  determines the maximum level of shop items
- if a mission has IO nodes as its target, the probability of a given IO node
  type being chosen as the target differs slightly from the Decker 1.12 values
- SAAB (Manufacturing) occurred twice in the list of target corporations
  for generic runs. The second entry was replaced by Mattel (Consumer).
- Zurich Orbital Gemeinschaft Bank in the list of target corporations
  for generic runs has been changed to Zuerich Orbital Gemeinschafts-Bank


THE LICENSE
===========
Everything in this .zip file is protected by the GNU General Public License 2.0 (GPL).
The license text is included in the download (filename : gpl.txt).



ROAD MAP
========

80 %
- the program can generate systems   70% done
- the matrix view
- player can do missions
- mission debriefing works


90 %
- sound is working
- options are available
- improvement : a table view for choosing matrix icons, support of custom sprites


version 1.9.1
- add system scripting to the mission scripting language
- the program can load and save games in its own format
- the program can import save games in the decker 1.9 format
- german localization


future tasks
- add syntax highlighting to the Script Checker
- the skin can be changed from within the program
- the free matrix
- the mission editor
- the mission editor can generate decker systems from FasTrix systems



FILES CONTAINED IN THIS DOWNLOAD
================================
The following files should be there :
decker.jar                     the executable game and the source code
readme.txt
gpl.txt                        the license text
ScriptChecker.bat              starts the script checker


changelog.txt
85%
2009-09-05
	ice action functions complete, except for GOING_HOME
2009-03-16
	double clicking a passive program in the system view no longer causes an error
2008-08-17
	minesweeper : added foxhound's graphics, updated the initial screen and first step of the tutorial
	hardcoded script function added : getTime()
	hardcoded script function renamed : getStructureStack() is now called getStack()
2008-08-07
	processing a user input event is now stopped when the displayed screen changes
	UIImage overhauled
	Battleships is playable (no restart or highscore table yet)
	added third party gif parser, to fix the gif animation problem in the next step
2008-08-06
	scripted function calls now put the structure the function is stored in on the stack for the function execution
	you can play a complete game of battleships
2008-08-04
	FunctionCall : if the supplied structure list contains the same structure twice in a row, it gets added to the structure stack only once
	delete(array,index) now calls the eventValueChanged() function
2008-08-01
	UIImage now reacts to angle changes and UIImage.<init> doesn't whine anymore if IMAGE.image is UNDEFINED
	step four of the battleships tutorial complete
	battleships place ships screen complete, final step without tutorial, because it would be too complicated
2008-07-31
	step three of the battleships tutorial complete
	button related mouse event handler functions now get told which button has changed its state
	fixed a bug in UIGenericComponent.eventValueChanged() that made it call DisplayedComponent.eventValueChanged() even if the changed varaible has nothing to do with displaying the generic component
	cleaned up some code
	removed all the standard messages during engine launch. set decker.model.Global.debug_level to something above 0 to make them reappear
2008-07-30
	added the ability to turn images in 90 degree steps
	continued work on battleships tutorial
2008-07-29
	started on the battleships tutorial
	changed the way localizations are marked, added the default_localization command
2008-07-26
	32 bit .bmps with no alpha channel (uppermost 8 bit always zero) are now interpreted as rrggbb images instead of completely transparent aarrggbb images
	instead of setting the initial_screen directly, Decker now calls the script function displayScreen() for it.
		this means the function initial_screen.setup() gets executed if it is defined
		and this change also ensures that the initial switching to initial_screen will always behave like all other screen switches, even if displayScreen() is changed in the future
2008-05-26
	the ice generator now works with the rearranged ice data
2008-05-17
	iceSoundAlarm finished
2008-05-16
	added iceSetResponseTargetNode
2008-04-23
	changed iceChooseAction a bit to solve the problems with ice staying hostile if it follows the player by chance
2008-03-19
	finished up iceChooseAction
2008-03-14
	rewrote the whole ice state change logic from scratch
2008-03-08
	fixed a number of bugs, added FOLLOWING ai behaviour
2008-03-05
   changed createSizedArray to initialize the array it creates with the value from the last parameter, if that value is not omitted (then it creates an empty array) and is not an integer or is below zero
2008-03-02
	ice should be able to move towards its target_node (the home node or the alarm node)
2008-03-01
	querying should qork completely
	a probe initiated a query for the first time :)
	ice should react correctly to alarms
	ice is now grouped into active, waiting and ungrouped passive ice, per area
2008-02-29
	removed the defunct & operator
	removed the "global value" language construct
2008-02-28
  recreated half the central ICE action function in script
2007-12-28
  the code for the default attack programs is mostly done
  the damage bar for the selected target works
  attacking works
2007-12-26
  double click event added
2007-12-23
  probe ice walks around
  you can mark ice

84%
2007-12-21
  the minimap works
2007-12-20
  !x now returns true for non-boolean values
  ice now has pretty much the same probability of being of a certain type as it had in Decker 1.12
  the two surplus COPs are back in and the area generator doesn't replace DATASTOREs with COPs anymore if there are only dummy IOs left and vice versa
  changed the default debug() depth to 3
  all ice is generated (the only things left to do are adding / marking mission goals and displaying ice)
  player and ice are displayed
2007-12-19
  changed the attack crash 9 ICE image from snake2 to snake3
  added all the ice images and data
2007-12-18
  added blocked regions to node types
  datastore nodes now have files
  the ? operators in a ? b : c ? d : e now execute left to right
  started to collect data on ICE
2007-12-17
  working on the system screen (where the action happens)
  program list displays programs (sorting and rowd ragging are not there yet)
  time display work
  can now generate whole systems (minus datastore files and ice)
  it's possible to walk around
2007-12-16
  system area generator completely translated from Java
  (it still needs to place ICE and portals, and to create whole systems)
2007-12-15
  started to convert the system area generator from Java
  added the % operator, and the hardcoded functions copyArraySection() and createSizedArray()
  fixed the problem where the function call f(x,) would lead to a parser error because of the omitted last parameter

83%
2007-12-15
  finished up the standard missions
2007-12-14
  mission descriptions are displayed, sorting the mission table works, details are automatically added to the mission descriptions
  generic missions work
2007-12-13
  UITextBlock displays everything but text
2007-12-12
  contract screen displays missions in the list
2007-12-10
  UIScrollpane glitch with maxed scrollbar position at the moment of content size reduction fixed

decker/model/Global.java
package decker.model;
import decker.util.*;
import decker.view.*;
import java.awt.*;
import java.io.*;
import java.util.Locale;
import java.util.Random;
import java.util.TreeSet;


/** contains three sections :
*      data methods and variables
*      view methods and variables
*      private methods
*/
public final class Global
{
	final static Random random = new Random();
	final static String COMMANDS = " if for while with print copy structure break constant localization default_localization "; // the string must start and end with spaces
	final static String BLOCK_INDENT = "   ";
	final static int PARSER_EXPRESSION_STACK_SIZE = 100;
	final static int DEFAULT_PRINT_DEPTH = 10;

	// id codes for hard coded functions. used by FunctionCall.executeFunctionCall() and Global.initializedataModel()
	final static int F_SIZE = 0, F_FILELIST = 1, F_SUBSTRING = 2, F_PIXELWIDTH = 3, F_PIXELHEIGHT = 4, F_EXIT_PROGRAM = 5, F_REPAINT = 6, F_INDEXOF = 7, F_IMAGE_EXISTS = 8, F_TO_LOWER_CASE = 9, F_TO_UPPER_CASE = 10, F_DATE_TEXT = 11, F_DEBUG = 12, F_INSERT = 13, F_RANDOM = 14, F_VALUE_TYPE = 15, F_DATE_DAY_OF_MONTH = 16, F_DATE_DAYS_IN_MONTH = 17, F_DELETE = 18, /*F_GET_STRUCTURE_STACK = 19,*/ F_IS_EXPANDABLE = 20, F_HAS_VARIABLE = 21, F_SCRIPT_NAME = 22, F_COPY_ARRAY_SECTION = 23, F_CREATE_SIZED_ARRAY = 24, F_INTEGER_TO_TEXT = 25, F_GET_STRUCTURE_MEMBER = 26, F_GET_STACK = 27, F_GET_TIME = 28;
	final static String[] FUNCTION_NAME = { "size", "filelist", "substring", "pixelwidth", "pixelheight", "exit_program", "repaint", "indexof", "image_exists", "toLowerCase", "toUpperCase", "date_text", "debug", "insert", "random", "value_type", "date_day_of_month", "date_days_in_month", "delete", /*"getStructureStack",*/ "isExpandable", "hasVariable", "scriptName", "copyArraySection", "createSizedArray", "integerToText", "getStructureMember", "printStack", "getTime" };

	public static int debug_level = 0;
	public static Locale[] accepted_locales = { Locale.getDefault(), new Locale("en") };
	public static Ruleset[] ruleset = new Ruleset[0];
	private final static Ruleset engine = new Ruleset("");
	private static Ruleset current_ruleset = new Ruleset("(dummy)");




//	final static void addStructureType (final StructureDefinition sd)  { current_ruleset.addStructureType(sd); }
public static Ruleset getCurrentRuleset ()  { return current_ruleset; }
	public static Structure getEngineData ()  { return ScriptNode.stack[ScriptNode.ENGINE_STACK_SLOT]; }



	/** sets things up for the game to launch and load the rulesets */
	public final static void initializeDataModel ()  {
		// set up the data stack
		engine.data.add("copyArraySection").set(new Function(F_COPY_ARRAY_SECTION, new String[]{ "from_array", "from_index", "to_array", "to_index", "entries" }));
		engine.data.add("createSizedArray").set(new Function(F_CREATE_SIZED_ARRAY, new String[]{ "size" }));
		engine.data.add("date_day_of_month").set(new Function(F_DATE_DAY_OF_MONTH, new String[]{ "year", "month", "day" }));
		engine.data.add("date_days_in_month").set(new Function(F_DATE_DAYS_IN_MONTH, new String[]{ "year", "month", "day" }));
		engine.data.add("date_text").set(new Function(F_DATE_TEXT, new String[]{ "year", "month", "day" }));
		engine.data.add("exit_program").set(new Function(F_EXIT_PROGRAM, new String[0]));
		engine.data.add("debug").set(new Function(F_DEBUG, new String[]{ "print_this", "to_console" }));
		engine.data.add("delete").set(new Function(F_DELETE, new String[]{ "array", "index" }));
		engine.data.add("filelist").set(new Function(F_FILELIST, new String[]{ "directory" }));
		engine.data.add("getStack").set(new Function(F_GET_STACK, new String[0]));
		engine.data.add("getTime").set(new Function(F_GET_TIME, new String[]{ "precision" }));
		engine.data.add("scriptName").set(new Function(F_SCRIPT_NAME, new String[0]));
		engine.data.add("getStructureMember").set(new Function(F_GET_STRUCTURE_MEMBER, new String[]{ "structure", "variable" }));
		engine.data.add("hasVariable").set(new Function(F_HAS_VARIABLE, new String[]{ "structure", "variable" }));
		engine.data.add("image_exists").set(new Function(F_IMAGE_EXISTS, new String[]{ "name" }));
		engine.data.add("indexof").set(new Function(F_INDEXOF, new String[]{ "what", "where", "direction", "start_at" }));
		engine.data.add("insert").set(new Function(F_INSERT, new String[]{ "array", "index" }));
		engine.data.add("integerToText").set(new Function(F_INTEGER_TO_TEXT, new String[]{ "integer", "radix", "min_digits" }));
		engine.data.add("isExpandable").set(new Function(F_IS_EXPANDABLE, new String[]{ "component" }));
		engine.data.add("pixelheight").set(new Function(F_PIXELHEIGHT, new String[]{ "component" }));
		engine.data.add("pixelwidth").set(new Function(F_PIXELWIDTH, new String[]{ "component" }));
		engine.data.add("random").set(new Function(F_RANDOM, new String[]{ "range_start", "range_end" }));
		engine.data.add("repaint").set(new Function(F_REPAINT, new String[0]));
		engine.data.add("size").set(new Function(F_SIZE, new String[]{ "thing" }));
		engine.data.add("substring").set(new Function(F_SUBSTRING, new String[]{ "string" , "from", "to" }));
		engine.data.add("toLowerCase").set(new Function(F_TO_LOWER_CASE, new String[]{ "text" }));
		engine.data.add("toUpperCase").set(new Function(F_TO_UPPER_CASE, new String[]{ "text" }));
		engine.data.add("value_type").set(new Function(F_VALUE_TYPE, new String[]{ "value" }));
		engine.data.add("displayed_screen").set(new Structure("VIEW", null)); // initialized with a dummy screen to avoid errors
		ScriptNode.stack[ScriptNode.ENGINE_STACK_SLOT] = engine.data;
		ScriptNode.stack[ScriptNode.RULESET_STACK_SLOT] = current_ruleset.data;
		ScriptNode.stack[ScriptNode.GLOBAL_STACK_SLOT] = new Structure("GLOBAL", null);
		ScriptNode.stack_size = ScriptNode.DEFAULT_GLOBAL_STACK_SIZE;
		ScriptNode.global_stack_size = ScriptNode.DEFAULT_GLOBAL_STACK_SIZE;
	}


	public final static void initializeRulesets ()  {
if (Global.debug_level > 0) {
System.out.println();
System.out.println(ruleset.length+" rulesets found");
System.out.println();
}
		final Ruleset r = current_ruleset;
		// first run the global scripts
if (Global.debug_level > 0)
System.out.println("running global scripts");
		setCurrentRuleset(engine);
		current_ruleset.initialize(accepted_locales);
		// then run the scripts of each ruleset
		for (int i = 0; i < ruleset.length; i++)  {
if (Global.debug_level > 0)
System.out.println("initializing ruleset "+ruleset[i].data.get("RULESET_NAME").toString());
			setCurrentRuleset(ruleset[i]);
			current_ruleset.initialize(accepted_locales);
		}
		setCurrentRuleset(r);
// print all ENGINE level structure types
// engine.data.get("STRUCTURE_TYPES").structure().print(System.err, "", true);
	}


	/** checks whether a given string contains an integer value */
	public final static boolean isInteger (final Object s)  {
		if (s != null && s instanceof String) {
			try {
				Integer.parseInt((String)s);
				return true;
			} catch (NumberFormatException ex) {}
		}
		return false;
	}


	/**
	 * Loads the scripts from the rulesets subfolder of the folder the jar sits in
	 */
	public static void loadRulesets ()  {
		File rulesetsFolder = new File("rulesets");
		if (!rulesetsFolder.exists() || !rulesetsFolder.isDirectory()) {
			throw new RuntimeException("Can not find 'rulesets' directory");
		}
			final File[] dir_list = rulesetsFolder.listFiles();
			bubblesort(dir_list);
			for (int d = 0; d < dir_list.length; d++) {
				if (dir_list[d].isDirectory() && !dir_list[d].getName().toLowerCase().endsWith(".svn")) {
if (Global.debug_level > 0)
System.out.println("loading Ruleset "+dir_list[d].getName());
					final Ruleset r = new Ruleset(dir_list[d].getName());
					final File[] script_list = dir_list[d].listFiles();
					bubblesort(script_list);
					// load the scripts from the ruleset
					boolean has_scripts = false;
					for (int i = 0; i < script_list.length; i++)
						if (script_list[i].getName().toLowerCase().endsWith(".txt")) {
							r.addScript(ScriptParser.parse(script_list[i], r));
							has_scripts = true;
						}
					// if the ruleset has at least one script, add it to the list of available rulesets
					if (has_scripts)
						ruleset = (Ruleset[]) ArrayModifier.addElement(ruleset, new Ruleset[ruleset.length+1], r);
				}
			}

			// add the global scripts which sit directly in the rulesets folder to the engine ruleset
if (Global.debug_level > 0)
System.out.println("loading engine scripts");
			for (int i = 0; i < dir_list.length; i++)
				if (dir_list[i].getName().toLowerCase().endsWith(".txt"))
					engine.addScript(ScriptParser.parse(dir_list[i], engine));
	}

	public static void setCurrentRuleset (final Ruleset r)  {
//final Ruleset old_ruleset = current_ruleset;
		current_ruleset = r;
		ScriptNode.stack[ScriptNode.RULESET_STACK_SLOT] = r.data;
// set the current screen or call ruleset.setup() or something
		// set all the ruleset specific values to their default values
		final Structure e = ScriptNode.stack[ScriptNode.ENGINE_STACK_SLOT];
		Value v;
		if ((v=e.get("displayed_screen")) != null)
			v.setConstant("UNDEFINED");
		if ((v=e.get("previous_displayed_screen"))!=null)
			v.set(new ArrayWrapper(new Value[0]));
		if ((v=e.get("DEFAULT_BORDER_THICKNESS"))!=null)
			v.set(2);
		if ((v=e.get("screen_overlays"))!=null)
			v.setConstant("UNDEFINED");
	}




// view methods ***************************************************************************************************************************************


	private final static ViewWrapper view_wrapper = new ViewWrapper();
	private static Component displayed_component;



	final static void displayTickerMessage (final String message) { view_wrapper.getView().displayTickerMessage(message); }


	public static Component getDisplayedComponent ()  { return displayed_component; }


	public static Value getDisplayedScreen ()  { final Structure e = ScriptNode.stack[ScriptNode.ENGINE_STACK_SLOT]; return (e==null) ? null : e.get("displayed_screen"); }


	public final static ViewWrapper getViewWrapper ()  { return view_wrapper; }


	public static void setDisplayedComponent (final Component c)  { displayed_component = c; }


// private methods ************************************************************************************************************************************


	private static void bubblesort (final File[] list)  {
		boolean list_modified = true;
		for(int i = 0; i+1 < list.length && list_modified; i++) {
			list_modified = false;
			for(int j = list.length-2; j >= i; j--) {
				if(list[j].getName().toLowerCase().compareTo(list[j+1].getName().toLowerCase()) > 0) {
					File x = list[j];
					list[j] = list[j+1];
					list[j+1] = x;
					list_modified = true;
				}
			}
		}
	}
}